src/data/dfinsupp/well_founded.lean

```diff
@@ -0,0 +1,227 @@
+type synonyms) carries a natural `(<)`.
+
+Notice that the definition of `dfinsupp.lex` says that `x < y` according to `dfinsupp.lex r s`
+iff there exists a coordinate `i : ι` such that `x i < y i` according to `s i`, and at all
+`r`-smaller coordinates `j` (i.e. satisfying `r j i`), `x` remains unchanged relative to `y`;
+in other words, coordinates `j` such that `¬ r j i` and `j ≠ i` are exactly where changes
+can happen arbitrarily. This explains the appearance of `rᶜ ⊓ (≠)` in
+`dfinsupp.acc_single` and `dfinsupp.well_founded`. When `r` is trichotomous (e.g. the `(<)`
+of a linear order), `¬ r j i ∧ j ≠ i` implies `r i j`, so it suffices to require `r.swap`
+to be well-founded.
+-/
+
+variables {ι : Type*} {α : ι → Type*}
+
+namespace dfinsupp
+
+variables [hz : Π i, has_zero (α i)] (r : ι → ι → Prop) (s : Π i, α i → α i → Prop)
+include hz
+
+open relation prod
+
+/-- This key lemma says that if a finitely supported dependent function `x₀` is obtained by merging
+  two such functions `x₁` and `x₂`, and if we evolve `x₀` down the `dfinsupp.lex` relation one
+  step and get `x`, we can always evolve one of `x₁` and `x₂` down the `dfinsupp.lex` relation
+  one step while keeping the other unchanged, and merge them back (possibly in a different way)
+  to get back `x`. In other words, the two parts evolve essentially independently under
+  `dfinsupp.lex`. This is used to show that a function `x` is accessible if
+  `dfinsupp.single i (x i)` is accessible for each `i` in the (finite) support of `x`
+  (`dfinsupp.lex.acc_of_single`). -/
+lemma lex_fibration : fibration
+  (inv_image (game_add (dfinsupp.lex r s) (dfinsupp.lex r s)) snd)
+  (dfinsupp.lex r s)
+  (λ x, merge x.1 x.2.1 x.2.2) :=
+begin
+  rintro ⟨p, x₁, x₂⟩ x ⟨i, hr, hs⟩,
+  simp_rw [merge_apply] at hs hr,
+  split_ifs at hs, classical,
+  work_on_goal 1
+  { refine ⟨⟨λ j, r j i → p j, merge (λ j, r j i) x₁ x, x₂⟩, game_add.fst ⟨i, _⟩, _⟩ },
+  work_on_goal 3
+  { refine ⟨⟨λ j, r j i ∧ p j, x₁, merge (λ j, r j i) x₂ x⟩, game_add.snd ⟨i, _⟩, _⟩ },
+  swap 3, iterate 2
+  { simp_rw merge_apply,
+    refine ⟨λ j h, if_pos h, _⟩,
+    convert hs,
+    refine ite_eq_right_iff.2 (λ h', (hr i h').symm ▸ _),
+    rw if_neg h <|> rw if_pos h },
+  all_goals { ext j, simp_rw merge_apply, split_ifs with h₁ h₂ },
+  { rw [hr j h₂, if_pos (h₁ h₂)] },
+  { refl },
+  { rw not_imp at h₁, rw [hr j h₁.1, if_neg h₁.2] },
+  { rw [hr j h₁.1, if_pos h₁.2] },
+  { rw [hr j h₂, if_neg (λ h', h₁ ⟨h₂, h'⟩)] },
+  { refl },
+end
+
+variables {r s}
+
+lemma lex.acc_of_single_erase [decidable_eq ι] {x : Π₀ i, α i} (i : ι)
+  (hs : acc (dfinsupp.lex r s) (single i (x i)))
+  (hu : acc (dfinsupp.lex r s) (x.erase i)) : acc (dfinsupp.lex r s) x :=
+begin
+  classical, rw ← merge_single_erase x i,
+  convert acc.of_fibration _ (lex_fibration r s) _,
+  work_on_goal 4 { refine ⟨_, _, _⟩ }, iterate 3 { refl },
+  exact inv_image.accessible snd (hs.prod_game_add hu),
```

Comment:
This proof looks hard to maintain. Could you make its structure a bit clearer?

Reply:
Isn't the proof short enough so maintenance won't be hard? I could maybe make it clearer by using `acc.of_fibration ... ⟨_, _, _⟩`; I'll look into that.

Reply:
golfed in [949d032](https://github.com/leanprover-community/mathlib/pull/16772/commits/949d032256dfb54be941868ec64bb79f9a2ce9dc)